add qnnpack path for hardtanh

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35779 add qnnpack path for hardtanh**

Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

vkuzo committed:
add qnnpack path for hardtanh

Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -14,12 +17,74 @@ namespace native {
+
+  initQNNPACK();
+
+  Tensor input_contig = input.contiguous();
```
We recently saw some perf implications of contiguous calls since it does contiguous in NCHW format. Can you use `input.suggest_memory_format()` here?

supriyar commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -14,12 +17,74 @@ namespace native {
+
+  Tensor qy = at::_empty_affine_quantized(
+    input_contig.sizes(),
+    input.options(),
```
Same here, can you specify the input memory format?

vkuzo replied:
looks like it's already coming through with the options

supriyar commented on test/test_quantized.py:
```diff
@@ -2705,6 +2705,32 @@ def test_mean(self, batch_size, channels, height, width, scale, zero_point):
+                                           dtype=torch_type)
+
+            qY_hat = torch.nn.quantized.functional.hardtanh(qX, min_val, max_val)
+            self.assertEqual(
```
Can we also compare the results against the server implementation? I think we can combine the test for server and mobile for this op.

vkuzo replied:
sure, we can reuse the body of the test, it probably does have to live in a separate wrapper though (check out the specs on `class TestQNNPackOps` for motivation)

vkuzo replied:
seems like only a few lines would be able to be reused, so left it as is for now (seemed overkill), but flexible - lmk!

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

z-a-f reviewed (approved):
LGTM

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]

vkuzo committed:
Update on "add qnnpack path for hardtanh"


Summary:

Adds a QNNPack path for the clamp kernel, which is useful for
hardtanh.

Test Plan:

python test/test_quantized.py TestQNNPackOps.test_hardtanh

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20778588](https://our.internmc.facebook.com/intern/diff/D20778588)

[ghstack-poisoned]
